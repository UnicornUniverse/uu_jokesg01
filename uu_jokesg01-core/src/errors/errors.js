// devkit-pure-exports
export * from "./get-error-lsi.js";
export * from "./get-error-status.js";
export * from "./property-error.js";
export * from "./uu-jokes-error.js";
export * from "./http-status.js";
