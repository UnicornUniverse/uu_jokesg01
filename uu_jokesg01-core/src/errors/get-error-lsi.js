//@@viewOn:imports
import HttpStatus from "./http-status";
import getErrorStatus from "./get-error-status";
import LsiData from "./errors-lsi";
//@@viewOff:imports

function getErrorLsi(errorData, customErrorLsi) {
  let lsi = getErrorMessage(errorData, customErrorLsi);

  if (!lsi) {
    const errorStatus = getErrorStatus(errorData);
    lsi = getErrorMessageByStatus(errorStatus, customErrorLsi);
  }

  return lsi;
}

function getErrorMessageByStatus(errorStatus, customErrorLsi) {
  let lsi;

  switch (errorStatus) {
    case HttpStatus.BaseNetworkError:
      lsi = customErrorLsi.baseNetworkError || LsiData.baseNetworkError;
      break;
    case HttpStatus.BadRequest:
      lsi = customErrorLsi.badRequest || LsiData.badRequest;
      break;
    case HttpStatus.Unauthorized:
      lsi = customErrorLsi.unauthorized || LsiData.unauthorized;
      break;
    case HttpStatus.Forbidden:
      lsi = customErrorLsi.forbidden || LsiData.forbidden;
      break;
    case HttpStatus.NotFound:
      lsi = customErrorLsi.notFound || LsiData.notFound;
      break;
    case HttpStatus.InternalServerError:
      lsi = customErrorLsi.internal || LsiData.internal;
      break;
    case HttpStatus.ServiceUnavailable:
      lsi = customErrorLsi.serviceUnavailable || LsiData.serviceUnavailable;
      break;
    case HttpStatus.GatewayTimeout:
      lsi = customErrorLsi.requestTimeout || LsiData.requestTimeout;
      break;
    default:
      lsi = customErrorLsi.defaultError || LsiData.defaultError;
  }

  return lsi;
}

function getErrorMessage(errorData, customErrorLsi) {
  const code = errorData?.error?.code || errorData.code;
  return customErrorLsi[code] || LsiData[code] || errorData?.error?.message;
}

//@@viewOn:exports
export { getErrorLsi };
export default getErrorLsi;
//@@viewOff:exports
