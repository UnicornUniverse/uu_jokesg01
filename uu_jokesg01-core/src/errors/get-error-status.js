export function getErrorStatus(errorData) {
  // note: there were cases when errorData without reparsing
  // were not behaving like an object
  // const errorData = JSON.parse(JSON.stringify(props.errorData));
  let status = errorData?.status;

  if (status === null || status === undefined) {
    return errorData?.error?.status;
  } else {
    return status;
  }
}

export default getErrorStatus;
