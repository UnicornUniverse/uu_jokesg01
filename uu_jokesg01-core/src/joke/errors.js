import Config from "./config/config";
import PropertyError from "../errors/property-error";

const ERROR_PREFIX = Config.TAG.toLowerCase().replaceAll(".", "-") + "provider/";

export class NoOidError extends PropertyError {
  static code = ERROR_PREFIX + "no-oid";

  constructor(cause) {
    super(NoOidError.code, "The required property oid is not defined!", cause);
  }
}

export default { NoOidError };
