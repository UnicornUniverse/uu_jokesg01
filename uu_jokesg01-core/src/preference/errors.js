import Config from "./config/config";
import PropertyError from "../errors/property-error";

const ERROR_PREFIX = Config.TAG.toLowerCase().replaceAll(".", "-") + "provider/";

export class NoUu5IdError extends PropertyError {
  static code = ERROR_PREFIX + "no-uu5-id";

  constructor(cause) {
    super(NoUu5IdError.code, "The required property Component identifier (uu5Id) is missing!", cause);
  }
}

export default { NoUu5IdError };
