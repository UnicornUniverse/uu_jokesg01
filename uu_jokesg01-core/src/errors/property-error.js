import { UuJokesError } from "../errors/uu-jokes-error";

export class PropertyError extends UuJokesError {
  constructor(code, message, cause) {
    super(code, message, cause);
  }
}

export default PropertyError;
