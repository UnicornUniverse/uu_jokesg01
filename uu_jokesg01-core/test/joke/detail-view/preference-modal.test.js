import { Test, VisualComponent, omitConsoleLogs } from "uu5g05-test";
import PreferenceModal from "../../../src/joke/detail-view/preference-modal";
import { UuJokesError } from "../../../src/errors/errors";
import queries from "../../queries/queries";
import enLsi from "../../../src/lsi/en.json";

const { screen } = Test;
const lsi = enLsi[PreferenceModal.uu5Tag];

function getDefaultProps() {
  const preferenceDataObject = {
    state: "ready",
    data: {
      showAuthor: true,
      showCategories: true,
      showCreationTime: true,
      disableUserPreferences: false,
    },
    handlerMap: {
      save: jest.fn(),
    },
  };

  return { preferenceDataObject, onSaveDone: jest.fn(), onCancel: jest.fn(), shown: true };
}

async function setup(props, options) {
  // ISSUE - uu5g05 - Uu5Forms.Form.View - invalid propTypes of Lsi
  // https://uuapp.plus4u.net/uu-sls-maing01/e80acdfaeb5d46748a04cfc7c10fdf4e/issueDetail?id=62f67d7f0b17bf002af36cfc
  omitConsoleLogs(
    "Warning: Failed prop type: Invalid prop `lsi` of type `string` supplied to `Uu5g05.Lsi`, expected `object`"
  );

  global.URL.createObjectURL = jest.fn();

  const { view, ...setupResult } = await VisualComponent.setup(
    PreferenceModal,
    { ...getDefaultProps(), ...props },
    { queries, ...options }
  );

  const submitBtn = screen.getByRole("button", { name: lsi.submit });
  const cancelBtn = screen.getByRole("button", { name: lsi.cancel });

  const inputs = {
    showAuthor: view.getFormsCheckbox(lsi.showAuthor),
    showCategories: view.getFormsCheckbox(lsi.showCategories),
    showCreationTime: view.getFormsCheckbox(lsi.showCreationTime),
  };

  return { ...setupResult, view, submitBtn, cancelBtn, inputs };
}

describe(`UuJokesCore.Joke.DetailView.PreferenceModal`, () => {
  it("submits form without changes", async () => {
    const { user, props, submitBtn } = await setup();

    await user.click(submitBtn);

    expect(props.preferenceDataObject.handlerMap.save).toHaveBeenCalledTimes(1);
    expect(props.preferenceDataObject.handlerMap.save.mock.lastCall[0]).toMatchSnapshot();
    expect(props.onSaveDone).toHaveBeenCalledTimes(1);
  });

  it("submits form with changes", async () => {
    const { user, props, submitBtn, inputs } = await setup();

    await user.click(inputs.showAuthor);
    await user.click(inputs.showCategories);
    await user.click(inputs.showCreationTime);
    await user.click(submitBtn);

    expect(props.preferenceDataObject.handlerMap.save).toHaveBeenCalledTimes(1);
    expect(props.preferenceDataObject.handlerMap.save.mock.lastCall[0]).toMatchSnapshot();
    expect(props.onSaveDone).toHaveBeenCalledTimes(1);
  });

  it("submits form with error", async () => {
    const { user, props, submitBtn } = await setup();

    PreferenceModal.logger.error = jest.fn();
    props.preferenceDataObject.handlerMap.save.mockImplementation(() => {
      throw new UuJokesError("baseNetworkError", "Test save error");
    });

    await user.click(submitBtn);

    //expect(screen.getByText(enLsi.Errors.baseError)).toBeVisible();
    expect(PreferenceModal.logger.error.mock.lastCall).toMatchSnapshot();
    expect(props.preferenceDataObject.handlerMap.save).toHaveBeenCalledTimes(1);
    expect(props.onSaveDone).toHaveBeenCalledTimes(0);
  });

  it("cancels form", async () => {
    const { user, props, cancelBtn } = await setup();

    await user.click(cancelBtn);

    expect(props.preferenceDataObject.handlerMap.save).toHaveBeenCalledTimes(0);
    expect(props.onCancel).toHaveBeenCalledTimes(1);
  });
});
