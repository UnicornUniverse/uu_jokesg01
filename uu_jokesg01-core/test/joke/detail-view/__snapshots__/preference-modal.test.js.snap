// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UuJokesCore.Joke.DetailView.PreferenceModal submits form with changes 1`] = `
Object {
  "showAuthor": false,
  "showCategories": false,
  "showCreationTime": false,
}
`;

exports[`UuJokesCore.Joke.DetailView.PreferenceModal submits form with error 1`] = `
Array [
  "Error while saving preference",
  [UuJokesError: Test save error],
]
`;

exports[`UuJokesCore.Joke.DetailView.PreferenceModal submits form without changes 1`] = `
Object {
  "showAuthor": true,
  "showCategories": true,
  "showCreationTime": true,
}
`;
